client/app/hearings/components/VirtualHearingModal.jsx

```
@@ -179,87 +195,94 @@ const TYPES = {
   }
 };
 
-class VirtualHearingModal extends React.Component {
-  constructor(props) {
-    super(props);
-
-    this.state = {
-      vetEmailError: null,
-      repEmailError: null
-    };
-
-    if (this.props.type === 'change_to_virtual') {
-      this.props.update({
-        veteranEmail: this.props.virtualHearing.veteranEmail || this.props.hearing.veteranEmailAddress,
-        representativeEmail:
-          this.props.virtualHearing.representativeEmail || this.props.hearing.representativeEmailAddress
-      });
-    }
-  }
-
-  validateForm = () => {
-    let { virtualHearing } = this.props;
-    let isValid = true;
+const VirtualHearingModal = (props) => {
+  const { closeModal, virtualHearing, reset, submit, type } = props;
+  const [vetEmailError, setVetEmailError] = useState(null);
+  const [repEmailError, setRepEmailError] = useState(null);
+  const [loading, setLoading] = useState(false);
+  const [success, setSuccess] = useState(false);
+  const typeSettings = TYPES[type];
 
+  const validateForm = () => {
     if (_.isEmpty(virtualHearing.veteranEmail)) {
-      this.setState({ vetEmailError: INVALID_EMAIL_FORMAT });
-      isValid = false;
+      setVetEmailError(INVALID_EMAIL_FORMAT);
+
+      return false;
     }
 
-    return isValid;
+    return true;
   };
 
-  onSubmit = () => {
-    let { submit } = this.props;
+  const onSubmit = () => {
+    if (!validateForm()) {
+      return;
+    }
+
+    setLoading(true);
 
-    if (this.validateForm()) {
-      submit().catch((error) => {
+    return submit()
+      ?.then(() => setSuccess(true))
```

`submit()` can return null if there were no changes to the form. This shouldn't ever happen in practice, so it's mostly a fail-safe.

---

Does safe-assign work on promises? Seems like this might throw an error if submit returns null instead of a promise that resolves to null or am I misunderstanding?

---

The safe-assign does work with promises, but it basically just leads to that whole submit function being a no-op (this is different than a promise that resolves to nothing, where the success callbacks to run).

---

I think both options are actually confusing for the user. Using safe-assign here basically causes the modal to be stuck in a loading state if `submit()` return null. That should never happen right now, but could if someone edits how the `submit()` prop is passed

---

Do you think we need to write a separate ticket to address handling null return values for the submit function?

---

I don't think we need one, but I'm open to creating one still if you think otherwise. Right now, I don't expect it to get into the "stuck loading" state I described after doing a bunch of manual testing.